File: fitrbackend/src/main/java/com/example/fitrbackend/dto/UserProfileResponse.java
package com.example.fitrbackend.dto;

import com.example.fitrbackend.model.ExperienceLevel;
import com.example.fitrbackend.model.Gender;
import com.example.fitrbackend.model.Goal;
import com.example.fitrbackend.model.Unit;
import java.time.Instant;
import lombok.Getter;
/**
 * Represents a response to a user profile request.
 */
public class UserProfileResponse {

    /**
     * The user's ID.
     */
    @Getter
    private final long id;

    /**
     * The user's ID.
     */
    @Getter
    private final long userId;

    /**
     * The user's first name.
     */
    @Getter
    private final String firstname;

    /**
     * The user's last name.
     */
    @Getter
    private final String lastname;

    /**
     * The user's email address.
     */
    @Getter
    private final String email;

    /**
     * The user's gender.
     */
    @Getter
    private final Gender gender;

    /**
     * The user's height.
     */
    @Getter
    private final float height;

    /**
     * The user's weight.
     */
    @Getter
    private final float weight;

    /**
     * The user's experience level.
     */
    @Getter
    private final ExperienceLevel experience;

    /**
     * The user's goal.
     */
    @Getter
    private final Goal goal;

    /**
     * The user's preferred unit of weight.
     */
    @Getter
    private final Unit preferredWeightUnit;

    /**
     * The user's preferred unit of distance.
     */
    @Getter
    private final Unit preferredDistanceUnit;

    @Getter
    private final Instant createdAt;

    /**
     * Creates a new instance of the {@link UserProfileResponse} class.
     *
     * @param id The user's ID.
     * @param userId The user's ID.
     * @param firstname The user's first name.
     * @param lastname The user's last name.
     * @param email The user's email address.
     * @param gender The user's gender.
     * @param height The user's height.
     * @param weight The user's weight.
     * @param experience The user's experience level.
     * @param goal The user's goal.
     * @param preferredWeightUnit The user's preferred unit of weight.
     * @param preferredDistanceUnit The user's preferred unit of distance.
     */
    public UserProfileResponse(long id, long userId, String firstname, String lastname, String email, Gender gender, float height, float weight, ExperienceLevel experience, Goal goal, Unit preferredWeightUnit, Unit preferredDistanceUnit, Instant createdAt) {
        this.id = id;
        this.userId = userId;
        this.firstname = firstname;
        this.lastname = lastname;
        this.email = email;
        this.gender = gender;
        this.height = height;
        this.weight = weight;
        this.experience = experience;
        this.goal = goal;
        this.preferredWeightUnit = preferredWeightUnit;
        this.preferredDistanceUnit = preferredDistanceUnit;
        this.createdAt = createdAt;
    }
}

File: fitrfrontend/fitrfrontend/Models/DTO/UserProfileResponse.swift
//
//  UserProfileResponse.swift
//  fitrfrontend
//
//  Created by Ambrose Blay on 2/17/26.
//

import Foundation

/// Represents a response to a user profile request.
struct UserProfileResponse: Codable, Identifiable {
    /// The profile ID
    let id: Int64
    
    /// The user's ID
    let userId: Int64
    
    /// The user's first name
    let firstname: String
    
    /// The user's last name
    let lastname: String
    
    /// The user's email address
    let email: String
    
    /// The user's gender
    let gender: Gender
    
    /// The user's height
    let height: Float
    
    /// The user's weight
    let weight: Float
    
    /// The user's experience level
    let experience: ExperienceLevel
    
    /// The user's goal
    let goal: Goal
    
    /// The user's preferred unit of weight
    let preferredWeightUnit: Unit
    
    /// The user's preferred unit of distance
    let preferredDistanceUnit: Unit
    
    /// Timestamp when the profile was created
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id, userId, firstname, lastname, email, gender, height, weight
        case experience, goal, preferredWeightUnit, preferredDistanceUnit, createdAt
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int64.self, forKey: .id)
        userId = try container.decode(Int64.self, forKey: .userId)
        firstname = try container.decode(String.self, forKey: .firstname)
        lastname = try container.decode(String.self, forKey: .lastname)
        email = try container.decode(String.self, forKey: .email)
        gender = try container.decode(Gender.self, forKey: .gender)
        height = try container.decode(Float.self, forKey: .height)
        weight = try container.decode(Float.self, forKey: .weight)
        experience = try container.decode(ExperienceLevel.self, forKey: .experience)
        goal = try container.decode(Goal.self, forKey: .goal)
        preferredWeightUnit = try container.decode(Unit.self, forKey: .preferredWeightUnit)
        preferredDistanceUnit = try container.decode(Unit.self, forKey: .preferredDistanceUnit)
        
        // Handle date decoding - supports both ISO 8601 string and timestamp
        let dateFormatter = ISO8601DateFormatter()
        if let dateString = try container.decodeIfPresent(String.self, forKey: .createdAt),
           let date = dateFormatter.date(from: dateString) {
            createdAt = date
        } else if let timestamp = try container.decodeIfPresent(Double.self, forKey: .createdAt) {
            createdAt = Date(timeIntervalSince1970: timestamp)
        } else {
            throw DecodingError.dataCorruptedError(forKey: .createdAt,
                                                   in: container,
                                                   debugDescription: "Cannot decode createdAt")
        }
    }
}
